go-controller/pkg/ovn/controller/services/services_controller.go

```diff
@@ -333,51 +352,18 @@ func (c *Controller) syncServices(key string) error {
 
 			// Node Port
 			if svcPort.NodePort != 0 {
-				gatewayRouters, _, err := gateway.GetOvnGateways()
-				if err != nil {
-					return errors.Wrapf(err, "failed to retrieve OVN gateway routers")
-				}
-				// Configure the NodePort in each Node Gateway Router
-				for _, gatewayRouter := range gatewayRouters {
-					lbID, err := gateway.GetGatewayLoadBalancer(gatewayRouter, svcPort.Protocol)
-					if err != nil {
-						klog.Warningf("Service Sync: Gateway router %s does not have load balancer (%v)",
-							gatewayRouter, err)
-						// TODO: why continue? should we error and requeue and retry?
-						continue
-					}
-					physicalIPs, err := gateway.GetGatewayPhysicalIPs(gatewayRouter)
-					if err != nil {
-						klog.Warningf("Service Sync: Gateway router %s does not have physical ips: %v",
-							gatewayRouter, err)
-						// TODO: why continue? should we error and requeue and retry?
-						continue
-					}
-					// With the physical_ip:port as the VIP, add an entry in 'load balancer'.
-					for _, physicalIP := range physicalIPs {
-						// only use the IPs of the same ClusterIP family
-						if utilnet.IsIPv6String(physicalIP) != utilnet.IsIPv6String(ip) {
-							continue
-						}
-						// VIP = NodeExternalIP:NodePort
-						vip := util.JoinHostPortInt32(physicalIP, svcPort.NodePort)
-						if c.needsOVNLBUpdate(eps, service) {
-							klog.V(4).Infof("Updating NodePort service %s/%s with VIP %s %s",
-								name, namespace, vip, svcPort.Protocol)
-							// Reconcile OVN, update the load balancer with current endpoints
-							if err := loadbalancer.UpdateLoadBalancer(lbID, vip, eps); err != nil {
-								c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToUpdateOVNLoadBalancer",
-									"Error trying to update OVN LoadBalancer for Service %s/%s: %v",
-									name, namespace, err)
-								return err
-							}
-							c.serviceTracker.setHasEndpoints(name, namespace)
-						}
-						c.serviceTracker.updateService(name, namespace, vip, svcPort.Protocol)
-						// mark the vip as processed
-						vipsTracked = vipsTracked.Delete(virtualIPKey(vip, svcPort.Protocol))
+				if c.needsOVNLBUpdate(eps.IPs, service) {
+					if err := createPerNodeVIPs(nil, svcPort.Protocol, svcPort.NodePort, eps.IPs, eps.Port); err != nil {
+						c.eventRecorder.Eventf(service, v1.EventTypeWarning, "FailedToUpdateOVNLoadBalancer",
+							"Error trying to update OVN LoadBalancer for Service %s/%s: %v",
+							name, namespace, err)
+						return err
 					}
+					c.serviceTracker.setHasEndpoints(name, namespace)
 				}
+				c.serviceTracker.updateService(name, namespace, vip, svcPort.Protocol)
```

Comment:
Here vip is the clusterip's vip but we are processing the nodeports (which have nodeip vips)

Reply:
nice catch! I could just make createPerNodeVIPs return all of the node vips, then iterate through them and updateService inside of the if c.needsOVNLBUpdate block. The caveat here is that the serviceTracker.updateService would not happen if ! c.needsOVNLBUpdate.

@aojea would that be ok or break something?

Reply:
I think it should be OK cause in service tracker we only ever check has endpoints, and if !c.needsOVNLBUpdate then there are no endpoints.

Reply:
actually line 366 has to change too to delete the right VIP, so perhaps i just have to rip off the bandaid and find all the physical ips separately.